MicroShift: merge kubeconfig to user defined KUBECONFIG or ~/.kube/config

praveenkumar opened the pull request:
With this patch, the kubeconfig file which is part of microshift is
merged with kubeconfig file which user provided as part of `KUBECONFIG`
env or default path (~/.kube/config) and set the default context to
microshift. This will allow user to start microshift and consume
directly without export the kubconfig path.

```
$ oc config get-contexts
CURRENT   NAME                                                                CLUSTER                                 AUTHINFO                                             NAMESPACE
          default/api-192-168-130-11-nip-io:6443/kubeadmin                    api-192-168-130-11-nip-io:6443          kubeadmin/api-192-168-130-11-nip-io:6443             default
          default/api-35-199-20-126-nip-io:6443/kubeadmin                     api-35-199-20-126-nip-io:6443           kubeadmin/api-35-199-20-126-nip-io:6443              default
          demo/api-192-168-130-11-nip-io:6443/kubeadmin                       api-192-168-130-11-nip-io:6443          kubeadmin/api-192-168-130-11-nip-io:6443             demo
*         microshift                                                          microshift                              user                                                 default
          test/api-192-168-130-11-nip-io:6443/kubeadmin                       api-192-168-130-11-nip-io:6443          kubeadmin/api-192-168-130-11-nip-io:6443             test

$ oc config current-context
microshift
```

praveenkumar committed:
MicroShift: merge kubeconfig to user defined KUBECONFIG or ~/.kube/config

With this patch, the kubeconfig file which is part of microshift is
merged with kubeconfig file which user provided as part of `KUBECONFIG`
env or default path (~/.kube/config) and set the default context to
microshift. This will allow user to start microshift and consume
directly without export the kubconfig path.

```
$ oc config get-contexts
CURRENT   NAME                                                                CLUSTER                                 AUTHINFO                                             NAMESPACE
          default/api-192-168-130-11-nip-io:6443/kubeadmin                    api-192-168-130-11-nip-io:6443          kubeadmin/api-192-168-130-11-nip-io:6443             default
          default/api-35-199-20-126-nip-io:6443/kubeadmin                     api-35-199-20-126-nip-io:6443           kubeadmin/api-35-199-20-126-nip-io:6443              default
          demo/api-192-168-130-11-nip-io:6443/kubeadmin                       api-192-168-130-11-nip-io:6443          kubeadmin/api-192-168-130-11-nip-io:6443             demo
*         microshift                                                          microshift                              user                                                 default
          test/api-192-168-130-11-nip-io:6443/kubeadmin                       api-192-168-130-11-nip-io:6443          kubeadmin/api-192-168-130-11-nip-io:6443             test

$ oc config current-context
microshift
```

praveenkumar commented:
This PR is important for podman desktop where it by default use `~/.kube/config` for getting the context.

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -87,6 +79,21 @@ func writeKubeconfig(ip string, clusterConfig *types.ClusterConfig) error {
+		return "", nil, err
+	}
+	// Make sure .kube/config exist if not then this will create
+	_, _ = os.OpenFile(kubeconfig, os.O_RDONLY|os.O_CREATE, 0600)
```
I'd write this as 
```
    kubeconfig := getGlobalKubeConfigPath()
    if !crcos.FileExists(kubeconfig) {
        dir := filepath.Dir(kubeconfig)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", nil, err
	}
	_, _ = os.OpenFile(kubeconfig, os.O_RDONLY|os.O_CREATE, 0600)
    }
    cfg, err := clientcmd.LoadFromFile(kubeconfig)
    if err != nil {
        return "", nil, err
    }
```
This makes it more obvious why mkdirall and OpenFile (I guess this could even go in a helper?) If you use OpenFile(), you need a maching `defer Close()`
k8s.io/client-go handles the same situation slightly differently, might be better to do the same?
```
func getConfigFromFile(filename string) (*clientcmdapi.Config, error) {
	config, err := LoadFromFile(filename)
	if err != nil && !os.IsNotExist(err) {
		return nil, err
	}
	if config == nil {
		config = clientcmdapi.NewConfig()
	}
	return config, nil
}
```

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -240,3 +247,34 @@ func updateServerDetailsToKubeConfig(input, output string) error {
 	return clientcmd.WriteToFile(*cfg, output)
 }
+
+func mergeKubeConfigFile(kubeConfigFile string) error {
```
Wondering if `ModifyConfig` in `vendor/k8s.io/client-go/tools/clientcmd/config.go` is achieving the same?

praveenkumar replied:
Following I tried but it is producing empty kubeconfig file at the end. I tried to find out if this is really used for merge the 2 different config but didn't get. Either I am doing something wrong or may be this is not the function for merging config (go-doc does specify it should do all but not sure how)
```golang
func mergeKubeConfigFile(kubeConfigFile string) error {
	_, kubeConf1, err := getGlobalKubeConfig()
	if err != nil {
		return err
	}
	kubeConf2, err := clientcmd.LoadFromFile(kubeConfigFile)
	if err != nil {
		return err
	}
	kubeConfigAccess1 := clientcmd.NewDefaultClientConfig(*kubeConf1, &clientcmd.ConfigOverrides{})
	return clientcmd.ModifyConfig(kubeConfigAccess1.ConfigAccess(), *kubeConf2, false)
}
```

praveenkumar replied:
https://github.com/crc-org/crc/issues/3602 created followup

praveenkumar committed:
cmd/Start: Remove the export kubeconfig in template

Since now kubconfig file is merged and default context is set
to microshift so you can start and directly consume api.
```
$ crc start
[...]
Started the MicroShift cluster.

Use the 'oc' command line interface:
  $ eval $(crc oc-env)
  $ oc COMMAND

$ oc get pods -A
NAMESPACE                  NAME                                  READY   STATUS             RESTARTS   AGE
openshift-dns              dns-default-29qgj                     2/2     Running            0          17s
openshift-dns              node-resolver-lmb5n                   1/1     Running            0          37s
openshift-ingress          router-default-8564985864-5rmzd       1/1     Running            0          31s
openshift-ovn-kubernetes   ovnkube-master-ghhss                  4/4     Running            0          37s
openshift-ovn-kubernetes   ovnkube-node-gjz5j                    1/1     Running            0          37s
openshift-service-ca       service-ca-54c746cbb7-mg9qw           1/1     Running            0          31s
openshift-storage          topolvm-controller-55b6f5f56b-64z6g   4/4     Running            0          37s
openshift-storage          topolvm-node-zscfc                    4/4     Running            0          17s
```

evidolob commented:
With this PR podman-desktop is able to find microshift context from kubeconfig file:

<img width="372" alt="Screenshot 2023-04-14 at 11 55 16" src="https://user-images.githubusercontent.com/929743/231996342-b4d6834e-7019-4176-8663-d40a8db05dfc.png">

praveenkumar commented:
Tested on M1 with cli and also checked podman desktop UI and it works as expected.
```
macos-arm64 git:(pr/3597) ✗ oc config get-contexts
CURRENT   NAME                                                                CLUSTER                                         AUTHINFO                                                NAMESPACE
          kind-kind                                                           kind-kind                                       kind-kind                                               
*         microshift                                                          microshift                                      user                                                    default
          prkumar-dev/api-sandbox-m2-ll9k-p1-openshiftapps-com:6443/prkumar   api-sandbox-m2-ll9k-p1-openshiftapps-com:6443   prkumar/api-sandbox-m2-ll9k-p1-openshiftapps-com:6443   prkumar-dev
          rancher-desktop                                                     rancher-desktop                                 rancher-desktop  
```
<img width="662" alt="Screenshot 2023-04-14 at 4 35 25 PM" src="https://user-images.githubusercontent.com/410007/232027896-49c6f9c8-bc4e-47eb-aeca-4c330f0ee7c9.png">

anjannath commented:
LGTM, working as expected, microshift context is added on `start` and removed on `crc delete`
```
❯ oc config get-contexts
CURRENT   NAME         CLUSTER      AUTHINFO   NAMESPACE
*         microshift   microshift   user       default
```

```
❯ crc delete -f
Deleted the instance

❯ cat ~/.kube/config
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users: null
```
